Add exception handling to win_domain_controller

agowa opened the pull request:
##### SUMMARY
Add error handling for two exceptions.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
win_domain_controller

##### ADDITIONAL INFORMATION

If a user specifies a domain to join that cannot be resolved by dns, the Install-ADDSDomainController command throws an Microsoft.ADRoles.Deployment.Common.Tests.TestFailedException exception.

If the module is either invoked twice without a reboot, or a reboot is pending for another reason the Install-ADDSDomainController command throws an Microsoft.DirectoryServices.Deployment.DCPromoExecutionException exception with Errorcode 15.

I added error handling for both of them, so that a user can handle them from within the playbook.

agowa committed:
Add exception handling to win_domain_controller

jborean93 commented on lib/ansible/modules/windows/win_domain_controller.ps1:
```diff
@@ -208,7 +208,30 @@ Try {
+                } catch {
+                    # We cannot directly catch Microsoft.ADRoles.Deployment.Common.Tests.TestFailedException
+                    # As the type is within an internal flagged class of Microsoft.ADRoles.Deployment.Common
+                    if (
```
I don't see a reason why we need this special handling here, when I test it locally the exception already shows the failed to resolve domain like message

```
Install-ADDSDomainController : Verification of user credential permissions failed. An Active Directory domain
controller for the domain "fake.domain.com" could not be contacted.
Ensure that you supplied the correct DNS domain name.
At line:1 char:1
+ Install-ADDSDomainController -DomainName 'fake.domain.com' -NoRebootO ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-ADDSDomainController], TestFailedException
    + FullyQualifiedErrorId : Test.VerifyUserCredentialPermissions.DCPromo.General.25,Microsoft.DirectoryServices.Depl
   oyment.PowerShell.Commands.InstallADDSDomainControllerCommand
```

It does have some extra fluff but I don't think we need to complicate the catch logic especially since this is an internal exception.

agowa replied:
I just did not want to do a catch all and this was the best solution I found to only match that particular exception. How could this be improved?
Or is it ok to do a catch all here?

jborean93 replied:
Honestly just don’t catch all as Ansible will now properly display the traceback and error message for you. You should only catch exceptions if the actual error is vague and you can improve it.

agowa replied:
But if I don't catch it the playbook fails. And exactly when this error is thrown that should not happen.

jborean93 replied:
Oh keep the first exception handler for DCPromoExecutionEngine but just remove the whole generic exception catch block.

agowa replied:
But than the user gets an error that cannot be handled within the playbook (and the whole play will fail) if Install-ADDSDomainController is unable to resolve the dns domain name... With the current exception handling a user could catch the error within the playbook and take action.

Maybe there is a misunderstanding:
> Ansible will now properly display the traceback and error message for you

The error is warped into a Fail-Json, so that ansible can not just display the error and error the whole play, but also to allow a user to work with the error using `ignore_error`, `register` and/or `retry`. And for that to work, ansible needs to know about this error.

jborean93 replied:
You can still pick up unhandled exceptions in a task, we even have a test for it. [This test](https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/tasks/main.yml#L28-L39) shows you what happens when you don't catch an explicit throw. The `msg` return value is the string `Unhandled exception while executing module: $($_.Exception.Message)` whereas the `exception` return value is the PowerShell error message and script stack trace.

Then [this test](https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/tasks/main.yml#L67-L78) shows you what happens when a .net function raises an unhandled exception. The same thing occurs and you can use `msg` or `exception` to parse.

Here is the module code that it is running to show you how the errors are fired https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_exec_wrapper/library/test_fail.ps1.

So for this case you would have something like

```
- win_domain_controller:
    dns_domain_name: unknown.domain
    domain_admin_user: username@unknown.domain
    domain_admin_password: password123!
    safe_mode_password: password123!
    state: domain_controller
  register: dc_result
  failed_when: dc_result is failed and dc_result.msg != "could not be contacted"
```

agowa replied:
I still don't fully agree to leave out that catch block, but that may be something for my internal fork and not something for general use. Also your solution may work (have to play with it a bit more) and also someone could do a dns lookup before trying to create a domain controller and catch that error before it even occurs. Thanks for the detailed explanation ;-)

Therefore I'm going to remove it for now.

jborean93 commented:
Also it would be great if you could add a changelog similar to https://github.com/ansible/ansible/pull/53480/files#diff-0b9ea3a63f047de3086e081e4eb479f2R3 for the module.

agowa committed:
Add changelog

agowa338 commented:
closing and reopening to trigger the ci

samdoran commented:
This needs to be rebased in order for tests to run properly.

agowa committed:
Fix PSUseDeclaredVarsMoreThanAssignments

agowa committed:
Remove dns domain cannot be resolved error message

As requested by PR review in https://github.com/ansible/ansible/pull/58234#discussion_r300509880

jborean93 commented:
rebuild_merge

jborean93 commented:
Looks like the bot is having issues, manually rerunning tests to ensure it isn't stale before merging.

jborean93 commented:
Feel free to backport this to stable-2.8 if you wish https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs.
